RPL border router sample application

rveerama1 opened the pull request:
This is a simple RPL based border router sample application. This can do routing between for example 802.15.4 and ethernet and act as a root node for RPL network. The router provides currently a very simple web UI for seeing the current nodes. Also new commands are added to net-shell to give better information about the RPL network. [Copied from https://github.com/zephyrproject-rtos/zephyr/pull/980].

Improve RPL neighbor and route handling functionality.
Improve net management event information mechanism.
Added more functionality in RPL Border Router sample to handle events and update web UI based on events, also draws network topology.
Added CoAP observer functionality to rpl_node sample.

jukkar committed:
samples: net: rpl: Simple RPL border router application

This is a very simple RPL border router sample application.
It provides HTTP(S) and net-shell interface for getting admin
information about neighbors and routes.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

dbkinder commented on include/net/websocket_console.h:
```diff
@@ -0,0 +1,60 @@
+
+/**
+ * @brief Websocket console library
+ * @defgroup websocket Websocket Console Library
```
include/net/websocket.h also defines a defgroup called ``websocket``.  Should this be named maybe ``websocket_console``

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,177 @@
+********
+
+The RPL border router sample application for Zephyr provides a HTTP(S) server
+and net shell for management purposes. Typically it would be used to connect to
```
Clarify what the "it" is that would be used, the net shell?

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,177 @@
+- Linux machine with web browser and the screen terminal emulator (optional).
+- Ethernet access for management purposes (optional).
+
+Note that there is no support for running RPL border router in QEMU, as the
```
change to "running an RPL border router"

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,177 @@
+- Ethernet access for management purposes (optional).
+
+Note that there is no support for running RPL border router in QEMU, as the
+border router requires access to a real radio network technology like
```
change "like" to "such as"

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,177 @@
+
+Note that there is no support for running RPL border router in QEMU, as the
+border router requires access to a real radio network technology like
+IEEE 802.15.4 which is not available in QEMU.
```
add a comma before "which"

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,177 @@
+border router requires access to a real radio network technology like
+IEEE 802.15.4 which is not available in QEMU.
+
+For testing purposes it is possible to compile the RPL border router for QEMU
```
change to:

    For testing purposes, you can compile ...

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,177 @@
+
+For testing purposes it is possible to compile the RPL border router for QEMU
+and do some testing with the web UI. But with QEMU, it is not possible to
+connect to RPL network and get information about the RPL nodes.
```
change to "connect to an RPL network"

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,177 @@
+	CONFIG_NET_APP_MY_IPV4_ADDR="192.0.2.1"
+
+Note that the IPv4 address is only used for connection to the integrated web
+server that provides admin web page for management purposes. The web server
```
change to "provides an admin"

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,177 @@
+Note that the IPv4 address is only used for connection to the integrated web
+server that provides admin web page for management purposes. The web server
+can also be connected using IPv6 address. If you do not have a web management
+network interface toward your host computer, then IPv4 support can be disabled
```
"toward" seems an odd word here.  Should it be "for"?

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,177 @@
+server that provides admin web page for management purposes. The web server
+can also be connected using IPv6 address. If you do not have a web management
+network interface toward your host computer, then IPv4 support can be disabled
+in configuration file by setting :option:`CONFIG_NET_IPV4` to "n".
```
change to "in the configuration file"

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,177 @@
+
+Note that the default project configuration file
+:file:`samples/net/rpl_border_router/prj.conf` does not currently provide
+a working system as there is no boards that would provide suitable network
```
either "is no board" or "are no boards"

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,177 @@
+interface support. The prj.conf file is provided only to compile test the
+border router sample application.
+
+It is possible to use the border router application with these boards and
```
Hmmm... I thought you just said there aren't any boards that are supported?

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,177 @@
+add-on cards:
+
+* FRDM-K64F with TI CC2520 card. See
+  https://wiki.zephyrproject.org/index.php?title=TI_CC2520
```
Nope, no such document there on our wiki any more.  @MaureenHelm do I need to recover that material from my wiki archive for you to recreate what's needed here?

dbkinder reviewed (changes requested):
suggested edits

rveerama1 committed:
samples: net: rpl: Add observer support to node application

If coap-client(e.g. RPL BR) wants to get notifications from
coap-server(RPL-node) about parent status, client has to register
observer for notifications.

RPL node application modified to toggle the LED. Do not maintain any
ON/OFF states. Also removed Sparrow (contiki border router) related
patch.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 committed:
samples: net: rpl: Enhance RPL border router application

Add CoAP client to RPL BR application. Which can be used to
control leds and get rpl related information from leaf nodes
though CoAP messages. Build toplogy matrics and send it to
Web UI.

Enhanced Web UI components (Interfaces, RPL, Neighbors, Routes
and Topology tabs). Added buttons to control leds on lead nodes.

Added JSON support to handle led on/off requests from UI.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 commented:
@dbkinder can please review doc updates?

dbkinder commented on samples/net/rpl_border_router/README.rst:
```diff
@@ -0,0 +1,174 @@
+It is possible to use the border router application with these boards and
+add-on cards:
+
+* FRDM-K64F with Freescale CR20A card.
```
This will render better like so:

    * `FRDM-K64F with Freescale CR20A card 
    <http://www.nxp.com/products/developer-resources/hardware-development-tools/freedom-development-boards/freedom-development-board-for-mcr20a-wireless-transceiver:FRDM-CR20A>`_.

dbkinder reviewed (changes requested):
one tweak

rveerama1 commented:
@dbkinder Fixed.

rveerama1 commented:
@dbkinder can you re-review this please?

dbkinder commented on samples/net/rpl_node/README.rst:
```diff
@@ -20,45 +20,19 @@ The sample exports the following resources through a CoAP server role:
-
-Wait until you see "Connected" message on console. Unset proxy in browser
-and open 127.0.0.1:8000.
+Follow the instructions from samples/net/rpl_border_router/README to run Zephyr
```
let's make this a live link with:

    Follow the instructions from :ref:`rpl-border-router-sample` to
    run the Zephyr RPL border router.

rveerama1 commented:
@dbkinder thanks, fixed comments, please take a look.

rveerama1 commented:
@dbkinder can you please take a look ?

rveerama1 commented:
@dbkinder can you please take a look ?

dbkinder reviewed (approved):
Thanks!

nashif commented on samples/net/rpl_node/prj_nrf52840_pca10056.conf:
```diff
@@ -0,0 +1,79 @@
+CONFIG_ARC_INIT=n
```
that does not belong here.

nashif reviewed (changes requested):
Please cleanup all the prj-*.conf files, create one that is global for platforms and then overlays for the specific boards and configurations. The configs have some duplicate configs, comments configs and probably obsolete ones as well, just keep those that are needed for the app to work. Take a look at samples/drivers/led_ws2812 for an example how to do that.

rveerama1 commented:
@nashif ping

nashif commented on samples/net/rpl_border_router/sample.yaml:
```diff
@@ -0,0 +1,14 @@
+    to be run in real device. This cannot be run in
+    QEMU.
+  name: RPL Border Router demo application
+tests:
```
please provide test entries for all the configs you are adding.

rveerama1 commented:
@nashif I update rpl_node sample according to your comments.
From rpl_border_router I dropped combination of frdm_k64f + cc2520 conf files. More info about  cc2520 + frdm combination is here https://github.com/zephyrproject-rtos/zephyr/issues/5274.
